### PR TITLE
Add all-gather coalescing for FSDP/ZeRO1

### PR DESCRIPTION
**Pull request by jeffhataws**
This PR adds all-gather coalescence support and use that in FSDP/ZeRO1 (replacing https://github.com/pytorch/xla/pull/5624). This PR is to be used in conjunction with https://github.com/openxla/xla/pull/5740 .

A separate and related PR for reduce-scatter coalescence that also enables using reduce-scatter's scale param in FSDP is https://github.com/pytorch/xla/pull/5938.

This is a revival of https://github.com/pytorch/xla/pull/4145 . Will need to address the comments.

### CODE REVIEW

**Comment by JackCaoG** [comment]
@jeffhataws let me know when you are done addressing comments, I will take another look

File: torch_xla/csrc/cross_replica_reduces.cpp [hunk 1]
```diff
@@ -309,6 +348,16 @@ ReduceScatterResult BuildReduceScatter(
   return {reduce_result, token_handler.GetNewToken(reduce_result)};
 }
 
+// moved from torch_xla/csrc/ops/all_reduce.cpp
```
Comment (JackCaoG) [inline torch_xla/csrc/cross_replica_reduces.cpp:351]: let's remove this comment and rename this to `GetOperandListWithToken`

File: torch_xla/csrc/cross_replica_reduces.h [hunk 2]
```diff
@@ -65,6 +71,11 @@ AllGatherResult BuildAllGather(xla::XlaOp input, xla::XlaOp token, int64_t dim,
                                const std::vector<std::vector<int64_t>>& groups,
                                bool pin_layout);
 
+AllGatherResultCoalesced BuildAllGatherCoalesced(
+    absl::Span<const xla::XlaOp>, xla::XlaOp token, int64_t dim,
```
Comment (JackCaoG) [inline torch_xla/csrc/cross_replica_reduces.h:75]: nit `absl::Span<const xla::XlaOp> inputs `

File: torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py [hunk 3]
```diff
@@ -295,6 +295,7 @@ def __init__(
       sharding_world_size: Optional[int] = None,
       shard_param_on_dim_0: bool = False,
       pin_layout_in_collective_ops: bool = True,
+      coalesce_all_gather_ops: bool = False,
```
Comment (JackCaoG) [inline torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py:298]: Do you mind explaining the change in this file? I think `coalesce_all_gather_ops ` is always `False` in our test, did you run into these issues with your own test?

File: torch_xla/csrc/cross_replica_reduces.cpp [hunk 4]
```diff
@@ -236,6 +236,45 @@ AllGatherResult BuildAllGather(xla::XlaOp input, xla::XlaOp token, int64_t dim,
...
+    bool pin_layout) {
+  std::vector<xla::ReplicaGroup> cc_groups = CreateReduceGroups(groups);
+  TokenHandler token_handler(token);
+  // TODO: We use pseudo-tokens ATM, which are real values. This need to be
+  // switched to use the real XLA Token once support has been added to XLA
+  // AllGather().
+  ReduceContext cc_ctx = GetReduceContext(inputs);
+  std::vector<xla::XlaOp> result(inputs.size());
+
+  for (auto& type_ctx : cc_ctx.contexts) {
```
Comment (JackCaoG) [inline torch_xla/csrc/cross_replica_reduces.cpp:251]: if you want to assume there is only one `type_ctx`, let's not use the for loop and `GetReduceContext` at all. This way we don't need to handle the token per type.

**Review by JackCaoG — COMMENTED** [review]
mostly lgtm beside the changes in FSDP. If we didn't change the default behavior of all-gather test should pass right?

I will look into reduce scatter one today, let's try to merge these two pr soon.

File: torch_xla/csrc/cross_replica_reduces.cpp (hunk 1 above)
Comment (jeffhataws) [reply torch_xla/csrc/cross_replica_reduces.cpp]: fixed

File: torch_xla/csrc/cross_replica_reduces.h (hunk 2 above)
Comment (jeffhataws) [reply torch_xla/csrc/cross_replica_reduces.h]: fixed

File: torch_xla/csrc/cross_replica_reduces.cpp (hunk 4 above)
Comment (jeffhataws) [reply torch_xla/csrc/cross_replica_reduces.cpp:251]: Let me check with others on this.

**Review by JackCaoG — APPROVED** [review]
Thanks! I think we should test allgather_coalesced using resnet on gpu to make sure we don't break it in the future. You can refer to existing test https://github.com/pytorch/xla/blob/2c4983dd1ab7600bd359740cb1f9f06c154d2933/.circleci/common.sh#L136.

we can do that in a separate pr.

**Pull request merged**